_posts/resources/2023-05-08-morello-linux-morelloie.md

```diff
@@ -0,0 +1,1096 @@
+
+```
+$ llvm-readelf -h hello
+...
+  Machine:                           AArch64
+  Entry point address:               0x210121
+...
+  Flags:                             0x10000, purecap
+...
+```
+
+Some output above is omitted. Here we may notice two things: the entry point address `0x210121` is odd (pun indented), and the flags field contains `0x10000` which means "purecap" (see [Morello ELF spec][ELF]). Each of these things tells us that this is a purecap binary.
+
+The `hello` binary you have now would run in the same way on any Morello platform, to run it on Morello IE, use:
+
+```
+$ morelloie -- ./hello
+Hello Morello
+```
+
+Double hyphen separates emulator’s parameters from the target application’s command line.
+
+Well done! We’ve run our first Morello app and it works! Surely, there are simpler ways to print "hello world". In fact, where is Morello in all this? Yes, we used a Morello toolchain to build the binary, but the code still looks the same as you would use for any other target platform. Well, that's the thing about Morello (and CHERI in general): normally, you shouldn't need to modify your code to compile it for a Morello target. Morello (and CHERI) "just works" behind the scenes gently taking care of your spatial memory safety.
+
+When compiling more involved examples, you may see new compiler warnings, usually they are there for a good reason and may indicate some memory safety issue. This is one of those situations when you may need to change your code to let it compile without warnings for a Morello target. However, you will notice as well that after this change the same code builds nicely for a non-Morello target, and it even may work better because a bug may have been fixed.
+
+## Purecap and hybrid
+
+Now it's time to talk about purecap (pure capability) and hybrid Morello applications. From an application developer's point of view, in a purecap program all pointers automatically become capabilities. This means that all memory accesses will happen via (and be controlled by) some specific capability. On the other hand, in a hybrid application, pointers are addresses unless explicitly defined as capabilities. So, a hybrid program would work in the same way as a native (AArch64) program until it encounters a pointer that is explicitly defined as a capability.
+
+Let's step back and consider these terms: "pointer", "address", and "capability". A pointer is an abstract token that gives you access to some memory. An address is the value of a pointer, i.e. a 64-bit integer that denotes the location in virtual memory. And a capability is one implementation of a pointer. We'll look deeper into capabilities and their properties in the next section. These three terms may be used in various contexts, but it's important to understand the difference between them whenever they happen to be in the same sentence.
```

Comment:
This use of the term “value” (although possibly accurate!) might be confusing -- since I guess in pure-capability C/C++, the value of a capability includes its metadata, but the integer cast or interpretation of it, is just the address? But I like the discussion of implementations of a pointers -- I wonder if it’s worth mentioning integer implementations here in the same language, so the reader thinks of them as being different implementations of the same idea?

Reply:
I will re-word this bit to avoid using the term "value" and instead will use "location [in virtual memory]".

Reply:
We have the word address, it means that, please just use it rather than some other clumsy non-standard alternative

Reply:
I'm using the word address except the place where I explain what "address" is.